src/lib/Monads.ml

```diff
@@ -127,60 +130,65 @@ struct
     M.scope @@ fun local ->
     {local with size = i + local.size}
 
-  let abort_if_inconsistent : 'a -> 'a m -> 'a m =
+  let abort_if_inconsistent : 'a m -> 'a m -> 'a m =
     fun abort m ->
     fun st ->
-    match CofEnv.consistency st.cof_env with
+    match CofEnv.Reduced.consistency st.cof_reduced_env with
     | `Consistent -> m st
-    | `Inconsistent -> M.ret abort st
+    | `Inconsistent -> abort st
 
   let lift_cmp (m : 'a compute) : 'a m =
-    fun {state; cof_env} ->
-    m {state; cof_env}
-
-  let replace_env cof_env m =
-    M.scope (fun local -> {local with cof_env}) @@
-    abort_if_inconsistent `Abort @@
-    let+ x = m in
-    `Ret x
-
-  let restrict phi m =
-    let* {cof_env} = M.read in
-    replace_env (CofEnv.assume cof_env phi) m
-
-  let left_invert_under_cofs phis m =
-    let* {cof_env} = M.read in
-    CM.left_invert_under_cofs cof_env phis @@
-    fun cof_env ->
-    let+ _ = replace_env cof_env m in ()
-
-  let left_invert_under_current_cof m = left_invert_under_cofs [] m
-
-  let bind_cof_proof phi m =
-    restrict phi @@ binder 1 m
+    fun {state; cof_reduced_env} ->
+    m {state; cof_env = CofEnv.Reduced.to_env cof_reduced_env}
+
+  let replace_env ~(abort : 'a m) (cof_reduced_env : CofEnv.reduced_env) (m : 'a m) : 'a m =
+    M.scope (fun local -> {local with cof_reduced_env}) @@
+    abort_if_inconsistent abort m
+
+  let restrict (type a) ~(splitter:(D.cof * a m) list -> a m) phis (m : a m) : a m =
+    let module P = struct
+      type t = a m
+      let zero = splitter []
+      let seq f cofs =
+        let l = List.map (fun cof -> cof , f cof) cofs in
+        splitter l
+    end in
```

Comment:
I will probably remove this scary `module` stuff once we have a correct version.